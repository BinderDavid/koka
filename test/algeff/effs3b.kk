import effs3

function mmap( f, xs ) {
  match(xs) {
    Nil -> Nil
    Cons(x,xx) -> Cons(f(x),mmap(f,xx))
  }
}

function xor() : amb bool {
  val p = flip() 
  val q = flip()
  (p||q) && not(p&&q)
}

function xorss() {
  val xs = mmap( fun(i) { xors() }, [1,2] )
  xs.join(",")
}

function upcast1() {
  val xs = mmap( id, ["x","y"]) // id is getting cps transformed
  if (flip()) then xs.join(",") else "darn"
}

function upcast2() {
  mmap(id,["x","y"]).join(",")  // mmap should use a default k, and id should be transformed...
}

function xors() : amb string {
  show(xor())
}
 
function amb_handle() {
  handler {
    return x -> [x]
    flip()   -> resume(False) + resume(True)
  }
}

