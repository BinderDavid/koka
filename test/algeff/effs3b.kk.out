False,True,True,False
.Op-flip : .op-flip
.Ops-flip : (flip : .op-flip) -> .ops-amb
.tag-amb : string
.tag-op-flip: string
amb_handle : forall<a,e> () -> ((action : () -> <amb|e> a) -> e list<a>)
flip : () -> amb bool
main : () -> console ()
mmap : forall<a,b,e> ((a) -> e b, list<a>) -> e list<b>
upcast1 : () -> amb string
upcast2 : () -> string
xor : () -> amb bool
xors : () -> amb string
xorss : () -> amb string